backend/regalloc/regalloc_rewrite.ml

```diff
@@ -37,16 +37,81 @@ type direction =
   | Load_after_list of Cfg.basic_instruction_list
   | Store_before_list of Cfg.basic_instruction_list
 
+(* Applies an optimization on the CFG outputted by [rewrite_gen] having one
+   temporary per variable per block rather than one per use of the variable,
+   reducing the number of spills and reloads needed for variables used multiple
+   times in a block. It iterates over each block and builds a substitution from
+   the first used temporary for each variable to all the other temporaries used
+   later for that variable, deleting now redundant reload/spill instructions
+   along the way. This currently does not support the case where spilled nodes
+   are used directly in instructions (if allowed by the ISA) without a new
+   temporary being created. This optimization does not use spills introduced in
+   inserted blocks due to spilling in the terminator to remove spills in the
+   original block. Returns the new block temporaries and removes the now block
+   temporaries and the instruction temporaries for now removed instructions from
+   [new_temporaries]. *)
```

Comment:
I'm not quite sure what the last two sentences mean:
- spills in terminators are untouched because their purpose is to reduce spills in the original block
- new block-temporaries are both removed from the block and returned, and this removes temps that were only used in removed instruction-level spill/reload pairs
Is that right?

Reply:
Spills in blocks inserted by `rewrite_gen` (due to spills in block terminators) do not share temporaries with the body of the original block and are hence not considered in this optimization.

Reply:
No new temporaries are created by this optimization. Some temporaries are promoted to block temporaries (and so moved from the list of new instruction temporaries to the list of new block temporaries). Instruction temporaries that are now redundant (due to being replaced by block temporaries) are removed from the list of new instruction temporaries.

Reply:
I have updated the comment using these explanations to hopefully make it clearer.